src/main/java/org/wildfly/security/auth/provider/oauth2/OAuth2Util.java

```diff
@@ -0,0 +1,194 @@
+            connection.setRequestProperty("Authorization", "Basic " + CodePointIterator.ofString(clientId + ":" + clientSecret).asUtf8().base64Encode().drainToString());
+
+            connection.getOutputStream().write(params);
+
+            return Json.createReader(new BufferedInputStream(connection.getInputStream())).readObject();
+        } catch (IOException ioe) {
+            if (connection != null && connection.getErrorStream() != null) {
+                InputStream errorStream = connection.getErrorStream();
+
+                try (BufferedReader reader = new BufferedReader(new InputStreamReader(errorStream))) {
+                    StringBuffer response = reader.lines().reduce(new StringBuffer(), StringBuffer::append, (buffer1, buffer2) -> buffer1);
+                    ElytronMessages.log.debugf(ioe, "Unexpected response from token introspection endpoint [%s]. Response: [%s]", tokenIntrospectionUrl, response);
+                } catch (IOException e) {
+                    throw ElytronMessages.log.oauth2RealmTokenIntrospectionFailed(ioe);
+                }
+            } else {
+                throw ElytronMessages.log.oauth2RealmTokenIntrospectionFailed(ioe);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Unknown error when invoking token introspection endpoint [" + tokenIntrospectionUrl + "].", e);
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns a {@link Attributes} instance based on the given {@link JsonObject}.
+     *
+     * @param claims a json object with the claims to extract
+     * @return an {@link Attributes} instance with attributes from the given json object
+     */
+    static Attributes toAttributes(JsonObject claims) {
+        return claims.entrySet().stream().reduce(new MapAttributes(), (mapAttributes, entry) -> {
+            String claimName = entry.getKey();
+            JsonValue claimValue = entry.getValue();
+
+            if (JsonValue.ValueType.ARRAY.equals(claimValue.getValueType())) {
+                JsonArray jsonArray = claims.getJsonArray(claimName);
+                jsonArray.forEach(arrayValue -> mapAttributes.addLast(claimName, asString(arrayValue)));
+            } else {
+                mapAttributes.addLast(claimName, asString(claimValue));
+            }
+
+            return mapAttributes;
+        }, (mapAttributes, mapAttributes2) -> mapAttributes);
+    }
+
+    private static String asString(JsonValue value) {
+        if (JsonValue.ValueType.STRING.equals(value.getValueType())) {
+            return ((JsonString) value).getString();
+        }
+
+        return value.toString();
+    }
+
+    private static HttpURLConnection openConnection(URL url) {
+        Assert.assertNotNull(url);
+        HttpURLConnection connection = null;
+
+        try {
+            connection = (HttpURLConnection) url.openConnection();
+
+            if (url.getProtocol().equalsIgnoreCase("https")) {
```

Comment:
I think we are going to need to be able to pass in the SSLContext to be used - I assume there is a MITM risk if our requests were intercepted?

Also I know this will fail as soon as it is run on a JVM running in FIPS mode.

From what you are using I think this can get away with having a Supplier for the SSLSocketFactory and it should also have a Supplier for the hostname verifier.